cuda_core/cuda/core/experimental/_memory/_pinned_memory_resource.pyx

```diff
@@ -0,0 +1,274 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from cuda.bindings cimport cydriver
+from cuda.core.experimental._memory._memory_pool cimport _MemPool, _MemPoolOptions
+from cuda.core.experimental._memory cimport _ipc
+from cuda.core.experimental._memory._ipc cimport IPCAllocationHandle
+from cuda.core.experimental._utils.cuda_utils cimport (
+    check_or_create_options,
+    HANDLE_RETURN,
+)
+
+from dataclasses import dataclass
+from typing import Optional
+import multiprocessing
+import platform  # no-cython-lint
+import subprocess
+import uuid
+import warnings
+
+from cuda.core.experimental._utils.cuda_utils import check_multiprocessing_start_method
+
+
+# Cache to ensure NUMA warning is only raised once per process
+cdef bint _numa_warning_shown = False
+
+
+def _check_numa_nodes():
+    """Check if system has multiple NUMA nodes and warn if so."""
+    global _numa_warning_shown
```

Comment:
Not a big deal since it's just a warning, but should we have some kind of lock around this to prevent multi-threading from possibly warning multiple times?

Reply:
In general, `cuda.core` has a few internal, module-level states that should be guarded by a lock. We haven't done that extensively. IIRC there should be an issue tracking the work for thread-safety, if not I'll create one.

Reply:
> IIRC there should be an issue tracking the work for thread-safety, if not I'll create one.

https://github.com/NVIDIA/cuda-python/issues/1389